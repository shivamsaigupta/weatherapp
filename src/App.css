.App {
  text-align: center;
  background-image: url('/imgs/bg.jpg');
  display: flex;
  height: 100vh;
}

.container{
  margin-top: 2rem;
  flex: 1;
}