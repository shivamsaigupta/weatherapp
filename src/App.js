import React from "react";
import "./App.css";
import Weather from "./components/weather.component";
import Search from "./components/search.component";
import {
  faSun,
  faCloud,
  faStarOfLife,
  faCloudShowersHeavy,
  faCloudRain,
  faBolt,
} from "@fortawesome/free-solid-svg-icons";

require("dotenv").config();

const API_KEY = process.env.REACT_APP_API_KEY;

class App extends React.Component {
  constructor() {
    super();
    this.state = {
      location: "",
      temp: 0,
      precip: "",
      icon: faSun,
      loading: true,
    };
  }

  componentDidMount() {
    this.getWeather("New York");
  }

  loadWeather = (e) => {
    e.preventDefault();
    const cityInput = e.target.elements.city.value;

    this.getWeather(cityInput);
  };

  getWeather = async (cityInput) => {
    const api_call = await fetch(
      `http://api.openweathermap.org/data/2.5/weather?q=${cityInput}&appid=${API_KEY}`
    );
    const res = await api_call.json();
    if (res.name !== undefined) {
      const tempCelsius = (res.main.temp - 273).toFixed(1);
      this.setState({
        location: res.name,
        temp: tempCelsius,
        precip: res.weather[0].main,
        icon: this.getIcon(res.weather[0].id),
        loading: false,
      });
    }
  };

  getIcon(value) {
    const thirdPlace = parseInt(value.toString()[0]);

    switch (thirdPlace) {
      case 2:
        return faBolt;
      case 3:
        return faCloudRain;
      case 5:
        return faCloudShowersHeavy;
      case 6:
        return faStarOfLife;
      case 7:
        return faCloud;
      case 8:
        return faCloud;
      default:
        return faSun;
    }
  }

  showSpinner() {
    return <div className="spinner-border" role="status"></div>;
  }

  render() {
    return (
      <div className="App">
        <div className="container">
          <Search onSubmit={this.loadWeather} />
          <Weather
            location={this.state.location}
            temp={this.state.temp}
            precip={this.state.precip}
            icon={this.state.icon}
            loading={this.state.loading}
          />
        </div>
      </div>
    );
  }
}

export default App;
