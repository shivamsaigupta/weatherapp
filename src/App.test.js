import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";

import "@testing-library/jest-dom";
import { getByTestId } from "@testing-library/dom";
import { render } from "@testing-library/react";

import App from "./App";

test("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<App />, div);
});

// doing the same mock testing with enzyme
// as well as react-testing-library
// for learning purposes

it("mock testing using enzyme", async (done) => {
  const mockResponse = {
    name: "Paris",
    main: {
      temp: "210",
    },
    weather: [
      {
        id: 700,
        main: "Cloudy",
      },
    ],
  };
  const mockJsonPromise = Promise.resolve(mockResponse);
  const mockFetchPromise = Promise.resolve({
    json: () => mockJsonPromise,
  });
  jest.spyOn(global, "fetch").mockImplementation(() => mockFetchPromise);
  const wrapper = shallow(<App />);

  expect(global.fetch).toHaveBeenCalledTimes(1);
  process.nextTick(() => {
    expect(wrapper).toMatchSnapshot();

    global.fetch.mockClear();
    done();
  });
});

// test("mock testing using react-testing-library", async (done) => {
//   const fakeRes = {
//     name: "Paris",
//     main: {
//       temp: "210",
//     },
//     weather: [
//       {
//         id: 700,
//         main: "Cloudy",
//       },
//     ],
//   };

//   const fetchMock = jest.spyOn(window, "fetch").mockImplementation(() => {
//     const fetchResponse = {
//       json: () => Promise.resolve(fakeRes),
//     };
//     return fetchResponse;
//   });

//   const div = document.createElement("div");
//   const { getByTestId } = render(<App />, div);

//   expect(fetchMock).toHaveBeenCalledTimes(1);

//   process.nextTick(() => {
//     expect(getByTestId("location")).toHaveTextContent("Paris");

//     window.fetch.mockRestore();
//     done();
//   });
// });
