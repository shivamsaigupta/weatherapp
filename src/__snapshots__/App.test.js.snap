// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches data from API when the server correctly responds 1`] = `
<div
  className="App"
>
  <div
    className="container"
  >
    <Weather
      location="Paris"
      precip="Cloudy"
      temp="210"
    />
  </div>
</div>
`;

exports[`mock testing using enzyme 1`] = `
<div
  className="App"
>
  <div
    className="container"
  >
    <Weather
      location="Paris"
      precip="Cloudy"
      temp="210"
    />
  </div>
</div>
`;
